src/plugins/controls/public/control_group/editor/control_editor.tsx

```diff
@@ -76,80 +90,103 @@ export const ControlEditor = ({
   getRelevantDataViewId,
   setLastUsedDataViewId,
 }: EditControlProps) => {
+  const { dataViews } = pluginServices.getHooks();
+  const { getIdsWithTitle, getDefaultId, get } = dataViews.useService();
+
   const { controls } = pluginServices.getServices();
   const { getControlTypes, getControlFactory } = controls;
+  const [state, setState] = useState<ControlEditorState>({
+    dataViewListItems: [],
+  });
 
-  const [selectedType, setSelectedType] = useState(
-    !isCreate && embeddable ? embeddable.type : getControlTypes()[0]
-  );
   const [defaultTitle, setDefaultTitle] = useState<string>();
   const [currentTitle, setCurrentTitle] = useState(title);
   const [currentWidth, setCurrentWidth] = useState(width);
   const [controlEditorValid, setControlEditorValid] = useState(false);
   const [selectedField, setSelectedField] = useState<string | undefined>(
-    embeddable
-      ? (embeddable.getInput() as DataControlInput).fieldName // CLEAN THIS ONCE OTHER PR GETS IN
-      : undefined
+    embeddable ? embeddable.getInput().fieldName : undefined
   );
 
-  const getControlTypeEditor = (type: string) => {
-    const factory = getControlFactory(type);
-    const ControlTypeEditor = (factory as IEditableControlFactory).controlEditorComponent;
-    return ControlTypeEditor ? (
-      <ControlTypeEditor
-        getRelevantDataViewId={getRelevantDataViewId}
-        setLastUsedDataViewId={setLastUsedDataViewId}
-        onChange={onTypeEditorChange}
-        setValidState={setControlEditorValid}
-        initialInput={embeddable?.getInput()}
-        selectedField={selectedField}
-        setSelectedField={setSelectedField}
-        setDefaultTitle={(newDefaultTitle) => {
-          if (!currentTitle || currentTitle === defaultTitle) {
-            setCurrentTitle(newDefaultTitle);
-            updateTitle(newDefaultTitle);
-          }
-          setDefaultTitle(newDefaultTitle);
-        }}
-      />
-    ) : null;
+  const doubleLinkFields = (dataView: DataView) => {
```

Comment:
Instead of using this function here and keeping the `fieldRestry` as part of the editor state, you could make this more readable & performant by using a `useMemo` which recalculates the `compatiblefieldRegistry` any time the data view changes. Could be something like:

``` typescript
const fieldRegistry = useMemo(() => {
  // create and return double linked fields.
}, [state.selectedDataView]);
```

This would also make it unnecessary to calculate the compatible fields on mount and whenever the data view changes.

Reply:
Great point - I figured I should be using either `useMemo` or `useCallback` in some way here, but couldn't quite land on a good way of doing it. I implemented this in e95afacfe83e4ef55f89ac315c8d76e5e4e147b7 - much cleaner now 🎉